src/fosslight_util/_get_downloadable_url.py

```
@@ -826,9 +876,120 @@ def get_download_location_for_pypi(link):
+        if versions:
+            return versions[-1]
+    except Exception as ex:
+        logger.debug(f"Failed to read Maven metadata ({metadata_url}): {ex}")
+    return ""
+
+
+def _maven_sources_from_directory(version_dir: str, artifact_id: str, version: str) -> str:
+    """Parse a Maven version directory listing for a sources jar link."""
+    preferred = f"{artifact_id}-{version}-sources.jar"
+    try:
+        html = urlopen(f"{version_dir.rstrip('/')}/").read().decode("utf8")
```

Ruff (0.16.1)

[error] Audit URL open for permitted schemes. Allowing use of `file:` or custom schemes is often unexpected.

(S310)